### PR TITLE
Remove query for all invitation codes from ES export

### PR DESCRIPTION
**Pull request by MocanaAtBroad**
## Context
We were getting all the invitation codes for the study every time we exported to ES. Now we get the invitation codes for each participant as part of the study participant query.

Only concern here is the usage of `Participant` class. Looks like sometimes we care about invitation codes (and medical providers) and sometimes we don't. I spent some time ensuring that removing the invitation code query and moving the invitation codes into Participant class does not break anything, but I don't fully understand the use cases, so there is a small chance that I did not fully capture everything.

## Checklist

- [ X] I have labeled the type of changes involved using the `C-*` labels.
- [X ] I have assessed potential risks and labeled using the `R-*` labels.
- [ X] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ X] I have considered security and privacy, and added `I-*` labels as needed
- [ X] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [ X ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [X ] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ X] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/ParticipantDao.sql.stg [hunk 1]
```diff
@@ -55,13 +55,26 @@ select en.user_study_enrollment_id as enrollment_id,
...
+       i.contact_email as i_contact_email,
+       i.notes as i_notes
   from user_study_enrollment as en
   join user as u on u.user_id = en.user_id
   left join user_profile as up on up.user_id = u.user_id
   left join default_mailing_address as def_addr on def_addr.participant_user_id = u.user_id
   left join mailing_address as addr on def_addr.address_id = addr.address_id
   left join user_medical_provider as mp on mp.user_id = u.user_id and mp.umbrella_study_id = en.study_id
+  left join invitation as i on i.user_id = mp.user_id and i.study_id = en.study_id
+  left join invitation_type as it on i.invitation_type_id = it.invitation_type_id
```
Comment (yufengwng) [inline pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/ParticipantDao.sql.stg:77]: I'm a bit worried about adding a left join here. A participant has only one profile, one default address, and multiple medical providers. So the query here is meant to product a result set that effectively create a one-to-many relation as "one participant to many medical providers".

But if we add in this left join here, then it expands the result set to "one participant to many medical providers to many invitations", which might produce duplicate results.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/ParticipantDao.java [hunk 2]
```diff
@@ -137,10 +141,16 @@ public void accumulate(Map<Long, Participant> container, RowView row) {
                 }
                 return new Participant(status, user);
             });
-
-            if (row.getColumn("m_user_medical_provider_id", Long.class) != null) {
+            Long medicalProviderId = row.getColumn("m_user_medical_provider_id", Long.class);
+            if (medicalProviderId != null
+                    && participant.getProviders().stream().noneMatch(p -> p.getUserMedicalProviderId() == medicalProviderId)) {
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/ParticipantDao.java:146]: I see here in the reducer that we're dealing with potential duplicate results. But maybe it's cleaner to have a separate query?

**Review by yufengwng — APPROVED** [review]
If this works as expected, then I'm okay with pushing forward on this.

File: pepper-apis/src/test/java/org/broadinstitute/ddp/export/DataExporterTest.java [hunk 3]
```diff
@@ -209,7 +211,17 @@ public void testExtractParticipants() {
...
+            handle.attach(InvitationFactory.class)
+                    .createInvitation(InvitationType.RECRUITMENT, invitationCode1, testData.getStudyId(),
+                            testData.getUserId(), email1);
+
+            String invitationCode2 = UUID.randomUUID().toString();
+            String email2 = testData.getUserGuid() + "@" + invitationCode2;
+            handle.attach(InvitationFactory.class)
+                    .createInvitation(InvitationType.AGE_UP, invitationCode2, testData.getStudyId(),
+                            testData.getUserId(), email2);
             // Extract and test
```
Comment (MocanaAtBroad) [inline pepper-apis/src/test/java/org/broadinstitute/ddp/export/DataExporterTest.java:225]: @yufengwng Regarding your concern about the one to many relationship. I was also was worried about which is why I added this chunk of code. I am confident we are handling it correctly.
(This is one of the things I really don't like about JDBI: one to many handling is super klutzy and forces redundant code. But that is for another day)

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/InvitationDao.java [hunk 4]
```diff
@@ -33,12 +33,6 @@
     @RegisterConstructorMapper(InvitationDto.class)
     Optional<InvitationDto> findByInvitationGuid(@Bind("studyId") long studyId, @Bind("guid") String invitationGuid);
 
-    @SqlQuery("select i.*, it.invitation_type_code from invitation as i"
-            + "  join invitation_type as it on i.invitation_type_id = it.invitation_type_id"
-            + " where i.study_id = :studyId")
-    @RegisterConstructorMapper(InvitationDto.class)
-    List<InvitationDto> findAllInvitations(@Bind("studyId") long studyId);
-
     default void saveNotes(long invitationId, String notes) {
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/InvitationDao.java:36]: The nasty query. Bye-bye!

**Comment by yufengwng** [comment]
Running things locally, I have 64K invitations and loading that in takes up at least 60 MiB of memory, but when running this PR there is very little memory overhead. So this looks good to me.

**Comment by MocanaAtBroad** [comment]
> Running things locally, I have 64K invitations and loading that in takes up at least 60 MiB of memory, but when running this PR there is very little memory overhead. So this looks good to me.

**Pull request merged**